AppCenter/AppCenterTests/MSHttpClientTests.m

```diff
@@ -0,0 +1,676 @@
+
+  // When
+  [httpClient sendAsync:url
+                 method:method
+                headers:nil
+                   data:nil
+      completionHandler:^(NSData *responseBody, NSHTTPURLResponse *response, NSError *error) {
+        // Then
+        XCTAssertEqual(response.statusCode, MSHTTPCodesNo400BadRequest);
+        XCTAssertNotNil(responseBody);
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+
+  [self waitForExpectationsWithTimeout:kMSTestTimeout
+                               handler:^(NSError *_Nullable error) {
+                                 if (error) {
+                                   XCTFail(@"Expectation Failed with error: %@", error);
+                                 }
+                               }];
+  // Then
+  XCTAssertEqualObjects(actualRequest.URL, url);
+  XCTAssertEqualObjects(actualRequest.HTTPMethod, method);
+  XCTAssertEqual(numRequests, 1);
+}
+
+- (void)testRecoverableNSErrorRetriedWhenNetworkReturns {
+
+  // If
+  __block BOOL completionHandlerCalled = NO;
+  __block BOOL firstTime = YES;
+  __block NSURLRequest *actualRequest;
+  NSArray *retryIntervals = @[ @1, @2 ];
+  [OHHTTPStubs
+      stubRequestsPassingTest:^BOOL(__unused NSURLRequest *request) {
+        return YES;
+      }
+      withStubResponse:^OHHTTPStubsResponse *(NSURLRequest *request) {
+        actualRequest = request;
+        if (firstTime) {
+
+          // Simulate network outage mid-request
+          [self simulateReachabilityChangedNotification:NotReachable];
+          firstTime = NO;
+          NSError *error = [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorCannotLoadFromNetwork userInfo:nil];
+          return [OHHTTPStubsResponse responseWithError:error];
+        }
+        return [OHHTTPStubsResponse responseWithData:[NSData data] statusCode:MSHTTPCodesNo204NoContent headers:nil];
+      }];
+  __weak XCTestExpectation *expectation = [self expectationWithDescription:@""];
+  MSHttpClient *httpClient = [[MSHttpClient alloc] initWithMaxHttpConnectionsPerHost:nil
+                                                                      retryIntervals:retryIntervals
+                                                                        reachability:self.reachabilityMock];
+  NSURL *url = [NSURL URLWithString:@"https://mock/something?a=b"];
+  NSString *method = @"DELETE";
+
+  // When
+  [httpClient sendAsync:url
+                 method:method
+                headers:nil
+                   data:nil
+      completionHandler:^(__unused NSData *responseBody, __unused NSHTTPURLResponse *response, __unused NSError *error) {
+        completionHandlerCalled = YES;
+        XCTAssertNotNil(responseBody);
+        XCTAssertNotNil(response);
+        XCTAssertEqual(response.statusCode, MSHTTPCodesNo204NoContent);
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+
+  // Wait a little to ensure that the completion handler is not invoked yet.
+  sleep(1);
```

Comment:
Do you really need to sleep here? You can move L245-249 within expectation handler after L256.
It would be `else` case of the `if`.

Reply:
I don't think that would work. We need L245-249 to execute before the completion handler is invoked (in fact, unless line 249 occurs, the completion handler will not be invoked). Since the expectation is fulfilled in the completion handler, the expectation handler is too late for that code to execute.